doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,390 @@
+default documentation for the top-level record (named `row` for the key schema
+and `envelope` for the value schema). In the example above, Materialize will
+consider the following sources of documentation in order:
+
+  * A `DOC ON TYPE` comment naming `v`.
+  * A comment on materialized view `v`.
+
+Changes made to comments *after* the the sink is created will not be reflected
+in the Avro schema.
+
+### User-facing documentation
+
+The details in the [Design](#design) section above will need to be worked into
+the Avro sink documentation. In particular, we should be sure to precisely
+explain the order in which Materialize searches
+for sources of documentation.
+
+Note that we expect the `COMMENT` interface to be the most common method by
+which users add comments to their sink schemas. The `DOC ON` option is primarily
+useful for the implementation, so that a `CREATE SINK` command can be fully self
+describing. We do expect users to use `DOC ON` in a pinch, e.g. if a single sink
+needs to customize the documentation for a nested type, rather than pulling its
+default description from its comment.
+
+### Implementation
+
+#### Parser
+
+The extensions to the parser are straightforward.
+
+#### Planner
+
+The planner will "freeze" any comments that have been promoted to documentation
+by rewriting the `CREATE SINK` statement to contain those comments as explicit
+`DOC ON` options. This ensures that the Avro schema generated for a sink is
+deterministic—i.e., that future changes to comments on the columns and types
+used by the sink will *not* result in Materialize generating a different schema
+on restart.
+
+#### Catalog
+
+No changes to the catalog are required. The documentation will be automatically
+stored in the catalog via the addition of the `DOC ON` option, as the catalog
+stores the raw DDL statement for each sink.
+
+#### Avro schema generation
+
+The `interchange::json::build_row_schema_json` will need to be adjusted to
+take the requested documentation as a parameter. Loosely, it could grow a
+`docs: BTreeMap<String, String>` parameter, where the key represents the
+field specifier and the value represents the requested documentation for that
+field.
+
+The sink code in the coordinator would need to convert the `DOC ON` comments
+to this `BTreeMap`.
+
+## Alternatives
+
+### User-provided Avro schemas
+
+Materialize could instead allow users to *provide* an Avro schema with all
+required `doc` fields. For example, users could pre-publish the desired schema
+to the schema registry, and then Materialize could load that schema at `CREATE
+SINK` time.
+
+There are several issues with this approach:
+
+  * It is not ergonomic. The provided schema must exactly match the schema
+    Materialize generates, *except* for the `doc` fields. Minor errors in
+    constructing the schema (e.g., using a `long` where an `int` is required, or
+    ordering fields wrong) will result in hard to debug failures.
```

Comment:
One idea, which came up recently in another context, was to add some `avro_schema_for(<relation>)` function that would output our generated avro schema as a string.

That would mitigate this concern a bit, since users could just edit the generated schema instead of having to get the types right a priori. I don't think it helps with the other concern however.

Reply:
Yes, we had a similar request to see the schema without creating the sink https://github.com/MaterializeInc/materialize/issues/21661